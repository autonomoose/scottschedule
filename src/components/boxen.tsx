/*
    various boxes
*/
import React from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';

interface iCaptionBox {
    caption: string,
    color?: string,
    xpad?: string, // rem
}
// quick caption box
export const CaptionBox = (props: iCaptionBox) => {
    const { caption, color, xpad } = props;
    const wkColor = color || 'inherit';
    const wkXpad = xpad || '0.5rem';
    return (
      <Box px={wkXpad} mt={-1.75}>
        <Typography variant='caption' color={wkColor}
          sx={{ bgcolor:'background.paper', position: 'relative', zIndex: '2'}} >
          &nbsp;{caption}&nbsp;
        </Typography>
      </Box>
)}

