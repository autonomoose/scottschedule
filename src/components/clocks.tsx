// more clocks
//   - clocks must avoid hooks, clocks are conditionals
//   - define new clock and register it with widget

import React from 'react';

import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';

// use this as an outside interface to clocks
interface ClockWidgetProps {
  currClock: string,
  onComplete: (status: string) => void,
  currSched?: string,
  descrSched?: string,
}
export const ClockWidget = (props: ClockWidgetProps) => {
    return(
      <Box display='flex' flexDirection='column'>
      {
        {
          'digital1': <ClockDigital1 onComplete={props.onComplete} />,
          'digital2': <ClockDigital2 onComplete={props.onComplete} />,
        }[props.currClock]
      }
      </Box>
)}

interface ClockProps {
  onComplete: (status: string) => void,
}
export const ClockDigital2 = (props: ClockProps) => {
    return(
      <>
        <Button data-testid='change clock' sx={{margin: 0}} onClick={() => props.onComplete('scheduler')}>
          <Typography variant='h1' id='mainclock' data-testid='mainclock'
            sx={{fontSize:150, fontWeight: 600, padding: 0, margin: 0}}>00:00</Typography>
        </Button>
        <Typography variant='subtitle1' id='mainpm' data-testid='mainpm' sx={{fontSize:20, fontWeight: 600}}>
          PM
        </Typography>
        <Box mx={4} display='flex' justifyContent='space-between'>
          <Typography mx={1} variant='h4' id='maindate' data-testid='maindate'>Day, 01/01</Typography>
          <Button onClick={() => props.onComplete('close')}>Scheduler</Button>
        </Box>
      </>
) };

export const ClockDigital1 = (props: ClockProps) => {
    return(
      <>
        <Button data-testid='change clock' sx={{margin: 0, padding: 0}} onClick={() => props.onComplete('digital2')}>
          <Box display='flex' justifyContent='flex-start' alignItems='stretch'>
            <Box mt={1} display='flex' alignItems='flex-start'>
              <Typography id='mainpm' data-testid='mainpm'
               sx={{lineHeight:1, fontSize:20, fontWeight: 600}}>
                PM
              </Typography>
            </Box>
            <Typography id='compclock' data-testid='compclock'
             sx={{marginLeft: '-2px', letterSpacing: '-30px', lineHeight:.8, fontSize:230, fontWeight: 600, padding: 0, margin: 0}}>
               0
            </Typography>
            <Box ml={3} mt={1}>
              <Typography id='compminutes' data-testid='compminutes'
               sx={{letterSpacing: '-4px',lineHeight:.92, fontSize:150, fontWeight: 800, padding: 0, margin: 0}}>
                00
              </Typography>
              <Typography id='maindate' data-testid='maindate'
               sx={{ textTransform: 'none', lineHeight:1, fontSize:26, padding: 0, margin: 0}}>
                Day, 01/01
              </Typography>
            </Box>
           </Box>
        </Button>
        <Box mx={4} display='flex' justifyContent='flex-end'>
          <Button onClick={() => props.onComplete('digital1-color')}>Color</Button>
          <Button onClick={() => props.onComplete('close')}>Scheduler</Button>
        </Box>
      </>
) };



