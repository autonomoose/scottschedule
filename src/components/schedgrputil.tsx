// sched and groups utilities and components
// exports default DisplaySchedGroup (test -base)
//    fetchSchedGroupsDB - full groups and schedules(test -base)
//  Group components CreateGroup (test -create),
//                     ModifyGroup (test -modify,-modempty)
//                     ChoiceSchedGroup (test -choice)
//  Schedule components ManSched (test -mansched)
//  ConnectTask (test -connev)

import React, { useEffect, useState } from 'react';
import { API } from 'aws-amplify';
import { useForm } from "react-hook-form";

import Autocomplete from '@mui/material/Autocomplete';
import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import IconButton from '@mui/material/IconButton';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import MenuItem from '@mui/material/MenuItem';
import TextField from '@mui/material/TextField';
import Typography from '@mui/material/Typography';

import { listSchedGroupsFull, iSchedGroupListDB } from '../graphql/queries';
import { mutAddEvents, mutDelEvents, mutAddRules, mutAddScheds } from '../graphql/mutations';

// -------------------------------------------------
interface CreateGroupProps {
  onComplete?: (status: string) => void,
  open: boolean
}
const mockComplete = (msg: string) => {console.log(msg)};
export const CreateGroup = (props: CreateGroupProps) => {
    const funComplete = (props.onComplete) ? props.onComplete : mockComplete;
    // form states
    const { register, handleSubmit, reset, formState } = useForm({
        defaultValues: {
            name: '',
            descr: '',
        }
    });
    const { isDirty, errors } = formState;

    interface FormNewGroupParms {
        name: string,
        descr: string,
    };
    const formNewGroupSubmit = async (data: FormNewGroupParms) => {
        try {
            const xdata = {'input': {
                'etype': 'gs',
                'evnames': data.name+"!args",
                'descr': data.descr,
                }
            };
            await API.graphql({query: mutAddEvents, variables: xdata});
            funComplete(data.name);
        } catch (result) {
            console.warn('failed group update', result);
        }
    };

    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="newGroup" onSubmit={handleSubmit(formNewGroupSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>
              Add New Group
            </Typography>
            <IconButton size='small' color='error' onClick={() => funComplete('')}>X</IconButton>
          </Box>

          <Box><label>
            Name <input type="text" size={12} data-testid="nameInput"
             {...register('name', { required: true, pattern: /\S+/, maxLength:16 })}
             aria-invalid={errors.name ? "true" : "false"}
            />
          </label></Box>
          <Box><label> Description
            <input type="text" size={30} data-testid="descrInput"
             {...register('descr', { required: true, pattern: /\S+/, maxLength:30 })}
             aria-invalid={errors.descr ? "true" : "false"}
            />
          </label></Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" onClick={() => reset()} disabled={!isDirty}>Reset</Button>
            <Button size="small" variant="contained" color="primary" type="submit" disabled={!isDirty}>Save</Button>
          </Box>

          </form>
        </Box>
      </Card></Box>
) };
// -------------------------------------------------
interface ModifyGroupProps {
  group: string,
  groupSched: iSchedGroup,
  onComplete?: (status: string) => void,
  open: boolean,
}
export const ModifyGroup = (props: ModifyGroupProps) => {
    const funComplete = (props.onComplete) ? props.onComplete : mockComplete;
    const wkGroup = props.groupSched;
    const wkName = props.group;

    // form states
    const { register, handleSubmit, reset, formState } = useForm({
        defaultValues: {
            descr: '',
        }
    });
    const { isDirty, errors } = formState;

    useEffect(() => {
        const defaultValues = {
            descr: (wkGroup && wkGroup.descr)? wkGroup.descr : '',
        }
        reset(defaultValues);
    }, [wkGroup] );

    interface FormModGroupParms {
        descr: string,
    };
    const formModGroupSubmit = async (data: FormModGroupParms) => {
        try {
            const xdata = {'input': {
                'etype': 'gs',
                'evnames': wkName+"!args",
                'descr': data.descr,
                }
            };
            await API.graphql({query: mutAddEvents, variables: xdata});
            funComplete(wkName);
        } catch (result) {
            console.warn('failed group update', result);
        }
    };
    interface FormDelEventParms {
        cmd: string,
    };
    const formDelEvent = async (data: FormDelEventParms) => {
        try {
            const xdata = {'input': {
                'etype': 'gs',
                'evnames': wkName+"!"+data.cmd,
                }
            };
            await API.graphql({query: mutDelEvents, variables: xdata});
            funComplete(wkName);
        } catch (result) {
            console.warn('failed delete', result);
        }
    };

    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="modGroup" onSubmit={handleSubmit(formModGroupSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>Modify Group</Typography>
            <IconButton size='small' color='error' onClick={() => funComplete('')}>X</IconButton>
          </Box>
          <Box display='flex' alignItems='center'>
            {wkName}
            { (wkGroup && wkGroup.schedNames.length === 0) &&
            <IconButton data-testid='delete' size='small' color='error' onClick={() => formDelEvent({'cmd': 'args'})}>X</IconButton>
            }
          </Box>

          <Box><label>
            <input type="text" size={30} data-testid="descrInput"
             {...register('descr', { required: true, pattern: /\S+/, maxLength:30 })}
             aria-invalid={errors.descr ? "true" : "false"}
            />
          </label></Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" onClick={() => reset()} disabled={!isDirty}>Reset</Button>
            <Button size="small" variant="contained" color="primary" type="submit" disabled={!isDirty}>Save</Button>
          </Box>
          </form>
          { (wkGroup) &&
          <>
            <Box mb={1} display='flex' justifyContent='space-around'>
              <span>Schedules ({wkGroup.schedNames.length}) </span>
              <Button onClick={() => funComplete('_'+wkName+'!_NEW_')}  size="small" variant="outlined" color="primary">New Schedule</Button>
            </Box>
            <List disablePadding dense sx={{marginLeft: '1em'}}>
            {
              wkGroup.schedNames.map(schedule => {
                return(
                  <ListItem button key={schedule.schedName} onClick={() => funComplete('_'+wkName+'!'+schedule.schedName)}>
                      {schedule.schedName} - {schedule.descr}
                  </ListItem>
              ) } )
            }
            </List>
          </>
          }
        </Box>
      </Card></Box>
) };

// -------------------------------------------------
interface ManSchedProps {
  groupSchedName: string, // group!sched or group!_NEW_
  gSchedule: iSchedule,
  evList: string[],
  onComplete?: (status: string) => void,
  open: boolean
}
interface FormManSchedParms {
    schedName: string,
    descr: string,
    begins: string,
    buttonName: string,
    sound: string,
    soundrepeat: string,
    warn: string,
};

// manage schedule
export const ManSched = (props: ManSchedProps) => {
    const funComplete = (props.onComplete) ? props.onComplete : mockComplete;
    let wkWords = props.groupSchedName.split('!');
    const groupName = wkWords.shift() || '';
    const schedName = wkWords.shift() || '';
    const currSchedule = props.gSchedule;
    const [schedEv, setSchedEv] = useState('');
    // form states
    const formDefaultVal: FormManSchedParms = {
        schedName: '',
        descr: '',
        begins: 'now',
        buttonName: '_same_',
        sound: '_default_',
        soundrepeat: '0',
        warn: '_none_',
    };
    const { register, handleSubmit, reset, formState } = useForm({defaultValues: formDefaultVal});
    const { isDirty, errors } = formState;

    useEffect(() => {
        let defaultValues: FormManSchedParms = formDefaultVal;
        if (schedName && schedName !== '_NEW_' && currSchedule) {
            defaultValues['schedName'] = schedName;
            defaultValues['descr'] = currSchedule.descr || '';
            defaultValues['begins'] = currSchedule.begins || 'now';
            if (currSchedule.buttonName || currSchedule.buttonName === '') {
                defaultValues['buttonName'] = currSchedule.buttonName;
            }
            if (currSchedule.sound) {
                if (currSchedule.sound['name'] || currSchedule.sound['name'] === '') {
                    defaultValues['sound'] = currSchedule.sound.name;
                }
                if (currSchedule.sound['repeat']) {
                    defaultValues['soundrepeat'] = currSchedule.sound.repeat.toString();
                }
            }
            if ('warn' in currSchedule) {
                defaultValues['warn'] = '_default_';
                if (currSchedule.warn && currSchedule.warn.sound) {
                    defaultValues['warn'] = currSchedule.warn.sound.name || '_default';
                }
            }
        }

        reset(defaultValues);
    }, [schedName, currSchedule] );

    const formManSchedSubmit = async (data: FormManSchedParms) => {
        let wkRetNames = groupName+"!"+data.schedName;
        if (schedName !== '' && schedName !== '_NEW_') {
            wkRetNames = groupName+"!"+schedName;
        }
        const wkEvNames = wkRetNames+"!args";

        try {
            const xdata = {'input': {
                'etype': 'gs',
                'evnames': wkEvNames,
                'descr': data.descr,
                'begins': data.begins,
                'button': (data.buttonName === '_same_')? data.schedName: data.buttonName,
                'sound': data.sound,
                'soundrepeat': data.soundrepeat,
                'warn': data.warn,
                }
            };
            await API.graphql({query: mutAddScheds, variables: xdata});
            funComplete(wkRetNames);
        } catch (result) {
            console.warn('failed sched update', result);
        }
    };
    interface FormDelEventParms {
        cmd: string,
    };
    const formDelEvent = async (data: FormDelEventParms) => {
        try {
            const keyNames = groupName+"!"+schedName;
            const xdata = {'input': {
                'etype': 'gs',
                'evnames': keyNames+"!"+data.cmd,
                }
            };
            await API.graphql({query: mutDelEvents, variables: xdata});
            funComplete((data.cmd === 'args')? '_'+keyNames: keyNames);
        } catch (result) {
            console.warn('failed delete', result);
        }
    };
    const formCallback = (status: string) => {
        setSchedEv('');
        if (status[0] !== '_') {
            funComplete(status);
        }
    };

    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="manSched" onSubmit={handleSubmit(formManSchedSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>
              { (schedName && schedName !== '_NEW_')
                ? <span>Modify Schedules </span>
                : <span>Add Schedule ({groupName})</span>
              }
            </Typography>
            <IconButton data-testid='cancel' size='small' color='error' onClick={() => funComplete('')}>X</IconButton>
          </Box>

          <Box display='flex' justifyContent='space-between'>
            <Box display={(schedName && schedName !== '_NEW_')?'none':'flex'}><label>
              Name <input type="text" size={12} data-testid="nameInput"
               {...register('schedName', { required: true, pattern: /\S+/, maxLength:16 })}
               aria-invalid={errors.schedName ? "true" : "false"}
              />
            </label></Box>
            {(schedName && schedName !== '_NEW_') &&
              <Box>
                {schedName}
                { (currSchedule && currSchedule.schedTasks.length === 0) &&
                  <IconButton size='small' color='error' onClick={() => formDelEvent({'cmd': 'args'})}>X</IconButton>
                }

              </Box>
            }

            <Box><label>
              Button <input type="text" size={6} data-testid="buttonInput"
               {...register('buttonName', { maxLength:8 })}
               aria-invalid={errors.buttonName ? "true" : "false"}
              />
            </label></Box>
          </Box>

          <Box><label>
            <input type="text" size={30} data-testid="descrInput"
             {...register('descr', { required: true, pattern: /\S+/, maxLength:30 })}
             aria-invalid={errors.descr ? "true" : "false"}
            />
          </label></Box>

          <Box><label>
            Begins <textarea rows={1} cols={27} data-testid="beginsInput"
             {...register('begins', { required: true, pattern: /\S+/, maxLength:50 })}
             aria-invalid={errors.begins ? "true" : "false"}
            ></textarea>
          </label></Box>

          <Box><label>
            Sound <input type="text" size={10} data-testid="soundInput"
             {...register('sound', { pattern: /\S+/, maxLength:20 })}
             aria-invalid={errors.sound ? "true" : "false"}
            />
          </label></Box>

          <Box><label>
            Repeat <input type="text" size={2} data-testid="soundRepeatInput"
             {...register('soundrepeat', { pattern: /\S+/, maxLength:2 })}
             aria-invalid={errors.soundrepeat ? "true" : "false"}
            />
          </label></Box>

          <Box><label>
            Warn <input type="text" size={10} data-testid="warnInput"
             {...register('warn', { pattern: /\S+/, maxLength:20 })}
             aria-invalid={errors.warn ? "true" : "false"}
            />
          </label></Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" onClick={() => reset()} disabled={!isDirty}>Reset</Button>
            <Button size="small" variant="contained" color="primary" type="submit" disabled={!isDirty}>Save</Button>
          </Box>

          </form>
          { (currSchedule) &&
          <>
            <Box mb={1} display='flex' justifyContent='space-around'>
              <span>Events ({currSchedule.schedTasks.length}) </span>
              <Button onClick={() => setSchedEv(groupName+'!'+schedName)}  size="small" variant="outlined" color="primary">Add Event</Button>
            </Box>
            <ConnectTask evList={props.evList} schedName={schedEv} onComplete={formCallback} open={(schedEv !== '')} />
            {
              currSchedule.schedTasks.map(task => {
                return(
                  <Box mx={2} key={task.evTaskId}>
                    <IconButton data-testid={'dconn-'+task.evTaskId} size='small' color='error' onClick={() => formDelEvent({'cmd': task.evTaskId })}>X</IconButton>

                    {task.evTaskId}
                  </Box>
              ) } )
            }
          </>
          }

        </Box>
      </Card></Box>
) };

// -------------------------------------------------
interface ConnectTaskProps {
  schedName: string,
  onComplete: (status: string) => void,
  evList: string[],
  open: boolean,
}
export const ConnectTask = (props: ConnectTaskProps) => {
    // form states
    const { register, handleSubmit, reset, formState } = useForm({
        defaultValues: {
            taskid: '',
        }
    });
    const { errors } = formState;


    const formConnectTaskCancel = async () => {
        props.onComplete('_'+props.schedName);
    }
    interface FormConnectTaskParms {
        taskid: string,
    };
    const formConnectTaskSubmit = async (data: FormConnectTaskParms) => {
        try {
            const xdata = {'input': {
                'etype': 'gs',
                'evnames': props.schedName+'!'+data.taskid,
                }
            };
            await API.graphql({query: mutAddRules, variables: xdata});
            props.onComplete(props.schedName);

        } catch (result) {
            console.warn('failed update connection', result);
            props.onComplete('');
        }
    };
    return(
      <Box display={(props.open)?'block': 'none'}>
      <Card style={{marginTop: '3px', maxWidth: 350, minWidth: 350, flex: '1 1', background: '#FAFAFA',
       boxShadow: '-5px 5px 12px #888888', borderRadius: '0 0 5px 5px'}}>
        <Box mx={1}>
          <form key="connectTask" onSubmit={handleSubmit(formConnectTaskSubmit)}>
          <Box display="flex" justifyContent="space-between" alignItems="baseline">
            <Typography variant='h6'>
              Connect Event
            </Typography>
          </Box>

          <Box display="flex" justifyContent="center">
            <Autocomplete
              options={props.evList}
              id="taskid" data-testid="taskid"
              sx={{ width: 300 }}
              clearOnEscape clearOnBlur
              renderInput={(params) => (
                <TextField {...params}
                  label="Event Name"
                  variant="outlined"
                  {...register('taskid', { required: true,})}
                  aria-invalid={errors.taskid ? "true" : "false"}
                />
              )}
            />
          </Box>

          <Box mt={2} display='flex' justifyContent='flex-end'>
            <Button size="small" variant="outlined" color='error' onClick={() => formConnectTaskCancel()}>Cancel</Button>
            <Button size="small" variant="outlined" onClick={() => reset()}>Reset</Button>
            <Button size="small" variant="contained" type="submit">Save</Button>
          </Box>

          </form>
        </Box>
      </Card></Box>
) }

// -------------------------------------------------
interface ChoiceSchedGroupProps {
    schedGroupList: iSchedGroupList,
    currgroup: string,
    setgroup: (event: React.ChangeEvent<HTMLInputElement>) => void,
}
export const ChoiceSchedGroup = (props:ChoiceSchedGroupProps) => (
    <TextField margin="dense" type="text" variant="outlined" size="small"
      value={props.currgroup} onChange={(e: React.ChangeEvent<HTMLInputElement>) => {props.setgroup(e)} }
      label="Schedule Group" id="schedgroup" sx={{minWidth: 120}}
      inputProps={{'data-testid': 'schedgroup'}}
      select
    >
      {(props.schedGroupList)
        ? Object.keys(props.schedGroupList).map(item => {
          return(
            <MenuItem key={item} value={item}>{item}</MenuItem>
           )})
        : <MenuItem value='new'>new</MenuItem>
      }
    </TextField>
);
// -------------------------------------------------
interface DisplaySchedGroupProps {
  group: string,
  groupSched: iSchedGroup,
  select?: (group: string) => void,
}
const DisplaySchedGroup = (props: DisplaySchedGroupProps) => {
    const wkGroup = props.groupSched;
    const wkName = props.group;
    return(
      <Box mb={1} key={wkName} sx={{borderTop: '1px solid grey'}}>
        <List disablePadding>
          <ListItem button sx={{marginRight: '1em'}} onClick={() => {props.select?.(wkName);}}>
            {wkName} - {wkGroup.descr}
          </ListItem>
        <List disablePadding dense sx={{marginLeft: '1em'}}>
        {
          wkGroup.schedNames.map(schedule => {
            // console.log(schedule);
          return(
            <ListItem button key={schedule.schedName} onClick={() => {props.select?.('_'+wkName+'!'+schedule.schedName);}}>
              {schedule.schedName} - {schedule.descr}
            </ListItem>
          )})
        }
        </List>
        </List>
      </Box>
) };

// -------------------------------------------------
export const fetchSchedGroupsDB = async (): Promise<iSchedGroupList> => {
    try {
        const result: any = await API.graphql({query: listSchedGroupsFull})

        const compactEvents = result.data.listSchedGroups.items.reduce((resdict: iGrpSchedTask, item: iSchedGroupListDB) => {
            const evkeys = item.evnames.split('!');
            // group,sched,ev
            let wkGroup = evkeys[0];
            let wkSched = evkeys[1];

            if (wkSched !== 'args' && evkeys[2] && evkeys[2] !== 'args') {
                if (!resdict[wkGroup+"!"+wkSched]) {
                    resdict[wkGroup+"!"+wkSched] = [];
                }
                resdict[wkGroup+"!"+wkSched].push({evTaskId: evkeys[2]});
            }
            return resdict;
        }, {});

        // console.log("eventlist", compactEvents);
        const compactGroups = result.data.listSchedGroups.items.reduce((resdict: iSchedGroupList, item: iSchedGroupListDB) => {
            const evkeys = item.evnames.split('!');
            // group,sched,ev
            let wkGroup = evkeys[0];
            if (!resdict[wkGroup]) {
                // setup basic group
                resdict[wkGroup] = {descr: '', schedNames: []};
            }
            if (evkeys[1] === 'args') {
                // group args
                resdict[wkGroup].descr = (item.descr)? item.descr: '';
            } else if (evkeys[2] === 'args') {
                // schedule args
                const wkSched = evkeys[1];
                let schedArgs: iSchedule = {schedName: wkSched, schedTasks: []};
                let wkSound: iEvsSound = {};

                if (compactEvents[wkGroup+"!"+wkSched]) {
                    schedArgs.schedTasks = compactEvents[wkGroup+"!"+wkSched];
                } else {
                    schedArgs.schedTasks = [];
                }
                schedArgs.descr = (item.descr)? item.descr: '';
                if (item.begins) {
                    schedArgs.begins = item.begins;
                }
                if (item.chain) {
                    schedArgs.chain = item.chain;
                }
                if (item.button || item.button === '') {
                    schedArgs.buttonName = item.button;
                }
                if (item.sound || item.sound === '') {
                    if (item.sound !== '_default_') {
                        wkSound['name'] = item.sound;
                    }
                }
                if (item.soundrepeat && parseInt(item.soundrepeat, 10) !== 0) {
                    wkSound['repeat'] = parseInt(item.soundrepeat, 10);
                }
                if (item.warn || item.warn === '') {
                    if (item.warn !== '_none_') {
                        schedArgs['warn'] = {};
                    }
                }

                if (wkSound && Object.keys(wkSound).length > 0) {
                    schedArgs['sound'] = wkSound;
                }
                resdict[wkGroup].schedNames.push(schedArgs)
            }
            return resdict;
        }, {});

        return(compactGroups);
    } catch (result) {
        return({});
    }
};

export default DisplaySchedGroup
