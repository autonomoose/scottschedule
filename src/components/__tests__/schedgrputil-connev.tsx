import React from "react";
import { render, waitFor } from "@testing-library/react";
import userEvent from '@testing-library/user-event'

import { API } from 'aws-amplify';
import { ConnectTask } from '../../components/schedgrputil';
jest.mock('aws-amplify');

const mockCallback = jest.fn();

const mytest = <ConnectTask schedName='testgrp' onComplete={mockCallback} open={true} />
const mySetup = () => {
    const utils = render(mytest);
    const canButton = utils.getByRole('button', {name: /cancel/i});
    const resetButton = utils.getByRole('button', {name: /reset/i});
    const saveButton = utils.getByRole('button', {name: /save/i});
    const taskFld = utils.getByTestId('taskid');

    return {
        ...utils,
        canButton,
        resetButton,
        saveButton,
        taskFld,
    }
};


describe("schedgrputil - connect event", () => {
  it("renders snapshot correctly", () => {
    const {container} = render(mytest);

    expect(container.firstChild).toMatchSnapshot();
  });
  it("starts with buttons in correct status", () => {
    const utils = mySetup();

    expect(utils.canButton).toBeEnabled();
    expect(utils.resetButton).toBeDisabled();
    expect(utils.saveButton).toBeDisabled();
  });
  it("cancels with button", () => {
    const utils = mySetup();

    userEvent.click(utils.canButton);
    expect(mockCallback).toHaveBeenLastCalledWith('_testgrp');
  });

  it("enables reset and save after event name modification", async () => {
    const utils = mySetup();

    userEvent.type(utils.taskFld, 'newev');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    expect(utils.saveButton).toBeEnabled();
  });
  it("handles reset after event modification", async () => {
    const utils = mySetup();

    userEvent.type(utils.taskFld, 'newev');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    userEvent.click(utils.resetButton);
    await waitFor(() => {
      expect(utils.resetButton).toBeDisabled();
    });
  });

  it("handles graphql error on save", async () => {
    const consoleWarnFn = jest.spyOn(console, 'warn').mockImplementation(() => jest.fn());
    const prevAPIgraphql = API.graphql;
    API.graphql = jest.fn(() => Promise.reject('mockreject'));
    const utils = mySetup();

    userEvent.type(utils.taskFld, 'newev');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    expect(utils.saveButton).toBeEnabled();
    userEvent.click(utils.saveButton);

    await waitFor(() => {
      expect(consoleWarnFn).toHaveBeenCalledTimes(1);
    });
    API.graphql = prevAPIgraphql;
    consoleWarnFn.mockRestore();
  });
  it("handles save after name modification", async () => {
    const prevAPIgraphql = API.graphql;
    API.graphql = jest.fn(() => Promise.resolve({}));
    const utils = mySetup();

    userEvent.type(utils.taskFld, 'newev');
    await waitFor(() => {
      expect(utils.resetButton).toBeEnabled();
    });
    expect(utils.saveButton).toBeEnabled();
    userEvent.click(utils.saveButton);

    await waitFor(() => {
      expect(mockCallback).toHaveBeenLastCalledWith('testgrp');
    });
    API.graphql = prevAPIgraphql;
  });

});


