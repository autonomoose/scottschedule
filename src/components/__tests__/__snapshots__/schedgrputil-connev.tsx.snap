// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedgrputil - connev (connect event) renders snapshot correctly 1`] = `
<div
  class="MuiBox-root css-13o7eu2"
>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    style="margin-top: 3px; max-width: 350px; min-width: 350px; flex: 1 1; background: rgb(250, 250, 250); box-shadow: -5px 5px 12px #888888; border-radius: 0 0 5px 5px;"
  >
    <div
      class="MuiBox-root css-ibnnmb"
    >
      <form>
        <div
          class="MuiBox-root css-h9yv4t"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Connect Event
          </h6>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1dv0nny-MuiAutocomplete-root"
            data-testid="taskid"
          >
            <div
              aria-invalid="false"
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="taskid"
                id="taskid-label"
              >
                Event Name
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  id="taskid"
                  name="taskid"
                  role="combobox"
                  spellcheck="false"
                  type="text"
                  value=""
                />
                <div
                  class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Open"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Open"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ArrowDropDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Event Name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-10igwnb"
        >
          <button
            class="MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root  css-1ir7th6-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Cancel
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root Mui-disabled  css-1k23hlb-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Reset
          </button>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButtonBase-root Mui-disabled  css-11qr2p8-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            Save
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
